matchms/filtering/SpectrumProcessor.py

```diff
@@ -24,116 +23,106 @@ class SpectrumProcessor:
-            self.add_custom_filter(filter_function[0], filter_function[1])
-
-    def add_matchms_filter(self, filter_spec: Union[Tuple[str, Dict[str, any]], str]):
-        """
-        Add a filter to the processing pipeline.
-
-        Parameters
-        ----------
-        filter_spec : str or tuple
-            Name of the filter function to add, or a tuple where the first element is the name of the
-            filter function and the second element is a dictionary containing additional arguments for the function.
+        for filter_name in filters:
+            self.parse_and_add_filter(filter_name)
+
+    def parse_and_add_filter(self, filter_description: Union[Tuple[str, Dict[str, any]],
+                                                             str,
+                                                             Tuple[Callable, Dict[str, any]],
+                                                             Callable],
+                             filter_position: Optional[int] = None):
+        """Adds a filter, by parsing the different allowed inputs.
+
+        filter:
+            Allowed formats:
+            str (has to be a matchms function name)
+            (str, {str, any} (has to be a matchms function name, followed by parameters)
+            Callable (can be matchms filter or custom made filter)
+            Callable, {str, any} (the dict should be parameters.
+        filter_position:
+            If None: Matchms filters are automatically ordered.
+            Custom filters will be added at the end of the filter list.
+            If not None, the filter will be added to the given position in the filter order list.
         """
-        if isinstance(filter_spec, str):
-            if filter_spec not in FILTER_FUNCTION_NAMES:
-                raise ValueError(f"Unknown filter type: {filter_spec} Should be known filter name or function.")
-            new_filter_func = FILTER_FUNCTION_NAMES[filter_spec]
-        elif isinstance(filter_spec, (tuple, list)):
-            filter_name, filter_args = filter_spec
-            if filter_name not in FILTER_FUNCTION_NAMES:
-                raise ValueError(f"Unknown filter type: {filter_name} Should be known filter name or function.")
-            new_filter_func = partial(FILTER_FUNCTION_NAMES[filter_name], **filter_args)
-            new_filter_func.__name__ = FILTER_FUNCTION_NAMES[filter_name].__name__
+        filter_args = None
+        if isinstance(filter_description, (tuple, list)):
+            if len(filter_description) == 1:
+                filter_function = filter_description[0]
+            elif len(filter_description) == 2:
+                filter_function = filter_description[0]
+                filter_args = filter_description[1]
+            else:
+                raise ValueError("The filter_function should contain only two values, "
+                                 "the first should be string or callable and the second a dictionary with settings")
```

Comment:
maybe using typedefs or even a class would then also make constructing these things a lot easier and would get rid of a lot of code.

Reply:
I like your suggestion of making a separate class for FilterFunction. However, this filter function would still need exactly the same code in the init. So in that sense not really solving the complexity. The reason we need this code is that we allow for multiple formats as input. But this is actually our design choice and I think we could simply change that. I have created an issue about this: #571